# Name : Abdul Samad 

# Batch =09

# Course Instructor :Sir sajid majeed 

# Date :7/8/25

# Email:abdulsamads3414@gmail.com

# Import the Libraries

In [105]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Load HR The dataset 

In [106]:
df=pd.read_csv(r"D:\Ned Machine Learning Course\Assignments\Assigmnet logistic regression on hr data\HR_comma_sep.csv")
df.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

# Gather basic information about the Hr data

In [107]:
df.shape

(14999, 10)

In [108]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   Department             14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


In [109]:
df.isnull().sum()

satisfaction_level       0
last_evaluation          0
number_project           0
average_montly_hours     0
time_spend_company       0
Work_accident            0
left                     0
promotion_last_5years    0
Department               0
salary                   0
dtype: int64

In [110]:
df.duplicated().sum()

np.int64(3008)

3,008 rows contain duplicate this is 20 % of the total rows

In [111]:
df.drop_duplicates(inplace=True)

In [112]:
df.shape

(11991, 10)

In [113]:
df.describe()

satisfaction_level  last_evaluation  number_project  \
count        11991.000000     11991.000000    11991.000000   
mean             0.629658         0.716683        3.802852   
std              0.241070         0.168343        1.163238   
min              0.090000         0.360000        2.000000   
25%              0.480000         0.570000        3.000000   
50%              0.660000         0.720000        4.000000   
75%              0.820000         0.860000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          11991.000000        11991.000000   11991.000000  11991.000000   
mean             200.473522            3.364857       0.154282      0.166041   
std               48.727813            1.330240       0.361234      0.372133   
min               96.000000            2.000000       0.000000      0.000000   
25%              157.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              243.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           11991.000000  
mean                0.016929  
std                 0.129012  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

In [114]:
df.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'Department', 'salary'],
      dtype='object')

In [115]:
# Rename the columns
df = df.rename(columns={'time_spend_company': 'tenure'})
df.head(2)

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   

   tenure  Work_accident  left  promotion_last_5years Department  salary  
0       3              0     1                      0      sales     low  
1       6              0     1                      0      sales  medium

In [116]:
# finding the number of people left by the 1 and 0
people_left =df[df["left"]==1]
people_stay=df[df["left"]==0]
print(people_left.shape)
print(people_stay.shape)

(1991, 10)
(10000, 10)


In [117]:
# check the outliers in the dataset
import plotly.express as px
import plotly.graph_objects as go
# box plot of tenure
fig = px.box(df, x="tenure", y="salary", title="Box Plot of Tenure vs Salary")
fig.show()


The above box plot shows that there are some outliers in tenure column . We can use the following code to remove outliers from the tenure column:

In [118]:
# Number of rows containing outliers
# Compute the 25th percentile and 75th percentile
q1 = df['tenure'].quantile(0.25)
q3 = df['tenure'].quantile(0.75)

# Compute the interquartile range
iqr = q3 - q1

# Compute the lower and upper bounds
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr
print("Lower bound: ", lower_bound)
print("Upper bound: ", upper_bound)
print("Number of rows containing outliers: ", df[(df['tenure'] < lower_bound) |(df['tenure'] > upper_bound)].shape[0])

Lower bound:  1.5
Upper bound:  5.5
Number of rows containing outliers:  824


# Visualization 

In [119]:
# bar plot salary vs left by plotly express
fig = px.bar(df, x="salary", y="left", title="Bar Plot of Salary vs Left")
fig.show()


concluded that ration of low salary people is more than the high salary people in leaveing the company .

In [120]:
# department wise left
fig = px.bar(df, x="Department", y="left", title="Bar Plot of Department vs Left")
fig.show()

sales team ratio is more than managment people

In [121]:
# tenure vs left
fig = px.bar(df, x="tenure", y="left", title="Bar Plot of Tenure vs Left")
fig.show()

People left the company within 3 years

# Build the Logistic regression model 

In [122]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelEncoder

In [123]:
# encode categorical columns 
le = LabelEncoder()
df["Department"] =le.fit_transform(df["Department"])
df["salary"]=le.fit_transform(df["salary"])

In [124]:
df.head(2)

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   

   tenure  Work_accident  left  promotion_last_5years  Department  salary  
0       3              0     1                      0           7       1  
1       6              0     1                      0           7       2

In [125]:
df.describe()

satisfaction_level  last_evaluation  number_project  \
count        11991.000000     11991.000000    11991.000000   
mean             0.629658         0.716683        3.802852   
std              0.241070         0.168343        1.163238   
min              0.090000         0.360000        2.000000   
25%              0.480000         0.570000        3.000000   
50%              0.660000         0.720000        4.000000   
75%              0.820000         0.860000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours        tenure  Work_accident          left  \
count          11991.000000  11991.000000   11991.000000  11991.000000   
mean             200.473522      3.364857       0.154282      0.166041   
std               48.727813      1.330240       0.361234      0.372133   
min               96.000000      2.000000       0.000000      0.000000   
25%              157.000000      3.000000       0.000000      0.000000   
50%              200.000000      3.000000       0.000000      0.000000   
75%              243.000000      4.000000       0.000000      0.000000   
max              310.000000     10.000000       1.000000      1.000000   

       promotion_last_5years    Department        salary  
count           11991.000000  11991.000000  11991.000000  
mean                0.016929      5.871153      1.356184  
std                 0.129012      2.899022      0.628071  
min                 0.000000      0.000000      0.000000  
25%                 0.000000      4.000000      1.000000  
50%                 0.000000      7.000000      1.000000  
75%                 0.000000      8.000000      2.000000  
max                 1.000000      9.000000      2.000000

In [126]:
# need standrad scaling of the average_montly_hours
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df["average_montly_hours"] = scaler.fit_transform(df[["average_montly_hours"]])
df.head(2)

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2             -0.892208   
1                0.80             0.86               5              1.262709   

   tenure  Work_accident  left  promotion_last_5years  Department  salary  
0       3              0     1                      0           7       1  
1       6              0     1                      0           7       2

In [127]:
df.describe()

satisfaction_level  last_evaluation  number_project  \
count        11991.000000     11991.000000    11991.000000   
mean             0.629658         0.716683        3.802852   
std              0.241070         0.168343        1.163238   
min              0.090000         0.360000        2.000000   
25%              0.480000         0.570000        3.000000   
50%              0.660000         0.720000        4.000000   
75%              0.820000         0.860000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours        tenure  Work_accident          left  \
count          1.199100e+04  11991.000000   11991.000000  11991.000000   
mean          -2.441361e-16      3.364857       0.154282      0.166041   
std            1.000042e+00      1.330240       0.361234      0.372133   
min           -2.144112e+00      2.000000       0.000000      0.000000   
25%           -8.922078e-01      3.000000       0.000000      0.000000   
50%           -9.718096e-03      3.000000       0.000000      0.000000   
75%            8.727716e-01      4.000000       0.000000      0.000000   
max            2.247814e+00     10.000000       1.000000      1.000000   

       promotion_last_5years    Department        salary  
count           11991.000000  11991.000000  11991.000000  
mean                0.016929      5.871153      1.356184  
std                 0.129012      2.899022      0.628071  
min                 0.000000      0.000000      0.000000  
25%                 0.000000      4.000000      1.000000  
50%                 0.000000      7.000000      1.000000  
75%                 0.000000      8.000000      2.000000  
max                 1.000000      9.000000      2.000000

In [128]:
# Select feature columns 
X = df[["satisfaction_level", "last_evaluation", "number_project", "average_montly_hours", "tenure", "Work_accident", "promotion_last_5years", "Department", "salary"]]
# Select target column
y = df["left"]

In [129]:
X.shape

(11991, 9)

In [130]:
y

0        1
1        1
2        1
3        1
4        1
        ..
11995    0
11996    0
11997    0
11998    0
11999    0
Name: left, Length: 11991, dtype: int64

In [131]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.20,random_state=42)


In [132]:
# Train Model 
LR=LogisticRegression()
LR.fit(X_train,y_train)


LogisticRegression()

In [137]:
# model accuracy
from sklearn.metrics import accuracy_score,classification_report
# Predict
y_pred=LR.predict(X_test)
# Accuracy
accuracy = accuracy_score(y_test, y_pred)
print("Model Accuracy:", accuracy)

# Classification Report
print(classification_report(y_test, y_pred))

Model Accuracy: 0.8345143809920801
              precision    recall  f1-score   support

           0       0.86      0.96      0.91      1998
           1       0.51      0.19      0.28       401

    accuracy                           0.83      2399
   macro avg       0.68      0.58      0.59      2399
weighted avg       0.80      0.83      0.80      2399

